core/consensus/babe/primitives/src/digest.rs

```diff
@@ -72,21 +74,26 @@ impl Encode for BabePreDigest {
 			authority_index: self.authority_index,
 			slot_number: self.slot_number,
 		};
-		parity_codec::Encode::encode(&tmp)
+		codec::Encode::encode(&tmp)
 	}
 }
 
+#[cfg(feature = "std")]
+impl codec::EncodeLike for BabePreDigest {}
+
 #[cfg(feature = "std")]
 impl Decode for BabePreDigest {
-	fn decode<R: Input>(i: &mut R) -> Option<Self> {
+	fn decode<R: Input>(i: &mut R) -> Result<Self, Error> {
 		let RawBabePreDigest { vrf_output, vrf_proof, authority_index, slot_number } = Decode::decode(i)?;
 
 		// Verify (at compile time) that the sizes in babe_primitives are correct
 		let _: [u8; super::VRF_OUTPUT_LENGTH] = vrf_output;
 		let _: [u8; super::VRF_PROOF_LENGTH] = vrf_proof;
-		Some(BabePreDigest {
-			vrf_proof: VRFProof::from_bytes(&vrf_proof).ok()?,
-			vrf_output: VRFOutput::from_bytes(&vrf_output).ok()?,
+		Ok(BabePreDigest {
+			vrf_proof: VRFProof::from_bytes(&vrf_proof)
+				.map_err(|_| Error::from("Invalid VRFProof"))?,
```

Comment:
yes improved, conversion is from signature error is not full though if we want to put more information we need to have a Cow instead of a static str. but I feel like it is ok.